### PR TITLE
[TRTLLM-12982][feat] support multi item scoring in LLM.encode

### PR DESCRIPTION
**Pull request by ixlmar**
## Description

This exposes the functionality added in https://github.com/flashinfer-ai/flashinfer/pull/1015 via `LLM.encode` (#12801). SGLang has been supporting this since https://github.com/sgl-project/sglang/pull/10979.

Note: Commit e50c18ad1e1ebe72f22f082672fc03fe1e6dafe8 is being integrated via #14634 but is required for tests to pass.

## Test Coverage

PR includes new tests.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Release Notes

* **New Features**
  * Added multi-item scoring capability to the encode API, enabling evaluation of multiple candidate answers within a single batch request.
  * Added `batch_indexed_model_output` and `copy_logits_to_host` parameters to the encode method for flexible output formatting and device control.

* **Tests**
  * Added comprehensive test coverage for multi-item scoring scenarios, input validation, and API constraints.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14693?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tensorrt_llm/llmapi/llm.py [hunk 1]
```diff
@@ -861,22 +881,47 @@ def encode(
...
+            start_idx = 0
+            for i, seq_len in enumerate(seq_lens.tolist()):
+                end_idx = start_idx + seq_len
+                results.append(
+                    EncoderOutput(
+                        logits=logits[start_idx:end_idx],
+                        prompt_token_ids=token_ids_list[i],
+                        prompt=prompts[i],
+                    ))
+                start_idx = end_idx
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/llmapi/llm.py:935]: The analysis appears correct and the behavior is a result of the `if logits.dim() > 1` condition introduced in #12801. Was there a specific use-case for having this @tingyangk?

**Review by Funatiq — APPROVED** [review]
Haven't looked at the attention backends, but everything else looks fine.

**Comment by venkywonka** [comment]
"item" seems to be heavily overloaded, with connotations at different layers (multimodal item means different from multi item but reads very close to the reader's eye)... could "candidate" be a better use. 
i might be bikeshedding here, but just wanted to make readability better. 
@ixlmar open to thoughts!

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 2]
```diff
@@ -2,14 +2,16 @@
...
 from dataclasses import dataclass, field
-from typing import Any, Dict, Literal, Optional
+from typing import Any, Dict, Literal, Optional, TypeAlias, cast
 
 import flashinfer
 import torch
 from flashinfer.jit.core import check_cuda_arch
 from typing_extensions import Self
 
+from tensorrt_llm._torch.pyexecutor.sampling_utils import torch_multi_arange
```
Comment (brb-nv) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py:13]: Nit: This seems to be the opposite dependency flow (intuitively for me, `pyexecutor` depends on `attention_backend`). Do you think we can house `torch_multi_arange` elsewhere?

File: tensorrt_llm/llmapi/llm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tingyangk) [reply tensorrt_llm/llmapi/llm.py:935]: Oh yes the analysis is correct! There was not a specific use-case. I missed the special case there. I think we can go for the suggested fix. Thanks!

File: tests/unittest/api_stability/references/llm.yaml [hunk 3]
```diff
@@ -355,6 +355,12 @@ methods:
       model_kwargs:
         annotation: Any
         default: inspect._empty
+      batch_indexed_model_output:
+        annotation: bool
+        default: True
+      copy_logits_to_host:
```
Comment (brb-nv) [inline tests/unittest/api_stability/references/llm.yaml:364]: Should this default to false instead? Or do we always want to copy logits to host for this codepath?

**Review by venkywonka — APPROVED** [review]
lgtm from llmapi!

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 4]
```diff
@@ -1629,6 +1805,7 @@ def decode_forward(plan_params: PlanParams, out: torch.Tensor):
             effective_mask_data = attention_mask_data
             if (self.flashinfer_backend == "trtllm-gen"
                     and attention_mask_data is not None):
+                logger.warning("Falling back to causal attention")
```
Comment (brb-nv) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py]: Nit: Does this get printed on every forward call? If yes, should we warn once to avoid spam?

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 5]
```diff
@@ -45,10 +60,15 @@ class PlanParams:
 
     attention_mask_type: AttentionMaskType
     attention_mask_data: Optional[torch.Tensor] = None
+    multi_item_params: Optional[MultiItemParams] = None
     sm_scale: Optional[float] = None
     window_left: Optional[int] = None
 
 
+# NB: Some features (multi-item scoring) are only supported with paged KV-cache wrapper (page_size=1)
+_RaggedPrefillWrapper: TypeAlias = flashinfer.BatchPrefillWithRaggedKVCacheWrapper | flashinfer.BatchPrefillWithPagedKVCacheWrapper
```
Comment (brb-nv) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py]: Calling both `BatchPrefillWith**Ragged**KVCacheWrapper` and `BatchPrefillWith**Paged**KVCacheWrapper` to be **Ragged** seems confusing.

**Review by brb-nv — APPROVED** [review]
Non-blocking comments.

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 2 above)
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:13]: I can relocate it to `tensorrt_llm/_torch/utils.py` (nearest common ancestor) and also reshuffle the tests accordingly. However, I'd leave that for a separate, follow-up PR.

File: tests/unittest/api_stability/references/llm.yaml (hunk 3 above)
Comment (ixlmar) [reply tests/unittest/api_stability/references/llm.yaml:364]: I chose the default in order not to break the base-branch behavior. Any opinions on this @tingyangk ? In any case, since changing this would be a breaking change, I'd still be inclined to open an independent micro-PR for that.

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 4 above)
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py]: I am unsure about the "right" way of dealing with this: The current solution is noisy, OTOH `logger.warning_once` carries the risk of getting lost among other output. Personally, I'd prefer requests specifying a custom attention mask to fail rather than quietly compute something that ignores the mask. Ideally, such requests would be rejected during a validation step elsewhere... Do have an opinion on this @lfr-0531 ?

Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py]: Changed this to use `.warning_once` for now.

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 5 above)
Comment (ixlmar) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py]: I extended the comment and introduced a distinct type to hopefully make that clearer.

File: tensorrt_llm/llmapi/llm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/llmapi/llm.py:935]: Applied the suggested fix
